File: Practice/AOJ/DSL/DSL2_A/dsl2_a.php
<?php
  $nq = array_map("intVal", explode(" ", trim(fgets(STDIN))));
  $n = $nq[0];
  $q = $nq[1];

  # セグ木のノード数は「初めてn以上になる2のべき乗」*2-1
  # ノードをi, 木の最下部の列数をmとしたとき
  # i + m - 1が実際のノードということになる

  $m = 1;
  while($m < $n){
    $m *= 2;
  }
  $INF = pow(2, 31) - 1;
  $tree = array_fill(0, 2 * $m-1, $INF);

  function update($a, $b){
    global $tree, $n, $m;
    $a += $m - 1;
    $tree[$a] = $b;
    while($a > 0){
      $a = floor(($a - 1)/2);
      $tree[$a] = min($tree[2 * $a + 1], $tree[2 * $a + 2]);
    }
  }

  function find($a, $b, $l, $r, $k=0){
    global $tree, $n, $INF;
    if($r <= $a || $b <= $l){
      return $INF;
    }
    if($a<=$l && $r<=$b){
      return $tree[$k];
    }else{
      $lv = find($a, $b, $l, floor(($l + $r) / 2), 2 * $k + 1);
      $rv = find($a, $b, floor(($l + $r) / 2), $r, 2 * $k + 2);
      return min($lv, $rv);
    }
  }

  for($i=0; $i<$q; $i++){
    $query = array_map("intVal", explode(" ", trim(fgets(STDIN))));
    $x = $query[1];
    $y = $query[2];
    if($query[0] == 0){
      update($x, $y);
    }else{
      echo find($x, $y+1, 0, $m) . PHP_EOL;
      
    }
  }
?>

File: Practice/AOJ/DSL/DSL2_A/dsl2_a.go
package main

import (
	"fmt"
)

var n, q int

type SegmentTree struct {
	n, k       int
	base, tree []int
}

func (st *SegmentTree) init() {
	if st.n < 1 {
		return
	}
	sz := st.n

	// 配列baseをセグ木で表現する。2^k-1個の配列をセグ木として使う
	st.k = 1
	for st.k < sz {
		st.k *= 2
	}
	st.tree = make([]int, 2*st.k-1)
	// // 葉から初期化する 最下層はk-1個なので+k-1でアクセス可能
	// for i := 0; i < sz; i++ {
	// 	st.tree[i+k-1] = st.base[i]
	// }
	// // 葉から根にむかって更新
	// for i := n - 2; i >= 0; i-- {
	// 	st.tree[i] = min(st.tree[2*i+1], st.tree[2*i+2])
	// }

	// ... この問題だとbaseがなくて最初からすべて1<<31-1で埋めた感じになるのでアドホックに
	for i := 0; i < 2*st.k-1; i++ {
		st.tree[i] = 1<<31 - 1
	}
}

func (st *SegmentTree) update(i, x int) {
	// 更新処理は葉からさかのぼっていく
	// whileでやるのでここで最下段のノードにアクセスする
	i += st.k - 1
	st.tree[i] = x
	for i > 0 {
		i = (i - 1) / 2 // ひとつ上の段に移動
		st.tree[i] = min(st.tree[2*i+1], st.tree[2*i+2])
	}
}
func (st *SegmentTree) find(i, x int) int {
	// findは根からくだっていく
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Scan(&n)
	fmt.Scan(&q)
	st := SegmentTree{n: n}
	st.init()
	for i := 0; i < q; i++ {
		var query, p, q int
		fmt.Scan(&query)
		fmt.Scan(&p)
		fmt.Scan(&q)
		if query == 0 {

		} else {

		}
	}
}
